Publishing metrics for job summary

diptanu opened the pull request:
Fixes #3465 

Publishing the job metrics in the leader node since they need to be published only from one of the servers.

Going to add docs, and related things once the implementation looks good to folks.

@dadgar @chelseakomlo

chelseakomlo commented on nomad/leader.go:
```diff
@@ -519,6 +522,52 @@ func (s *Server) periodicUnblockFailedEvals(stopCh chan struct{}) {
+				}
+				summary := raw.(*structs.JobSummary)
+				for name, tgSummary := range summary.Summary {
+					metrics.SetGauge([]string{"nomad", "job_summary", summary.JobID, name, "queued"}, float32(tgSummary.Queued))
```
As of Nomad 0.7, metrics are published in a tagged format. See: 
- https://www.nomadproject.io/api/metrics.html
- https://www.nomadproject.io/docs/agent/configuration/telemetry.html#backwards_compatible_metrics

diptanu replied:
Yeah I saw that, but I wasn't sure what the tags are going to be in this case.

diptanu replied:
Any suggestions?

dadgar replied:
I think the job should be a tag.

chelseakomlo commented on nomad/leader.go:
```diff
@@ -519,6 +522,52 @@ func (s *Server) periodicUnblockFailedEvals(stopCh chan struct{}) {
+			ws := memdb.NewWatchSet()
+			iter, err := state.JobSummaries(ws)
+			if err != nil {
+				timer.Reset(s.config.StatsCollectionInterval)
```
Is it necessary to reset the timer in each case statement?

diptanu replied:
@chelseakomlo Since we are resetting the timer at the end of the loop, we need to reset it before doing a `continue` in the error branches.

chelseakomlo replied:
If the timing doesn't need to be exact, could you reset the timer at the beginning of the loop for all cases? Otherwise then maybe add comments to explain why timers are set only for some branches.

diptanu replied:
@chelseakomlo I was just thinking since the time taken by the work is not O(1) but O(n) we might cascade a few ticks if the iteration takes longer for some reason like GC pauses or underlying OS issues. I can add comments to explain this.

dadgar replied:
I think it is worth the simpler code to just reset once at the top of the case

diptanu commented:
@chelseakomlo I added support for tagged metrics, currently making job id and taskgroup name as tags. Please let me know if that addresses your comment regarding that?

diptanu commented:
```
{
Labels: {
job: "example",
task_group: "cache"
},
Name: "nomad.nomad.job_summary.complete",
Value: 0
},
{
Labels: {
job: "example",
task_group: "cache"
},
Name: "nomad.nomad.job_summary.failed",
Value: 0
},
{
Labels: {
job: "example",
task_group: "cache"
},
Name: "nomad.nomad.job_summary.lost",
Value: 0
},
{
Labels: {
job: "example",
task_group: "cache"
},
Name: "nomad.nomad.job_summary.queued",
Value: 0
},
{
Labels: {
job: "example",
task_group: "cache"
},
Name: "nomad.nomad.job_summary.running",
Value: 1
},
{
Labels: {
job: "example",
task_group: "cache"
},
Name: "nomad.nomad.job_summary.starting",
Value: 0
},
```

jippi commented on nomad/leader.go:
```diff
@@ -519,6 +522,74 @@ func (s *Server) periodicUnblockFailedEvals(stopCh chan struct{}) {
+						metrics.SetGaugeWithLabels([]string{"nomad", "job_summary", "lost"},
+							float32(tgSummary.Lost), labels)
+					}
+					if s.config.BackwardsCompatibleMetrics {
```
i like the irony of a new telemetry feature have a `BackwardsCompatibleMetrics` check :P

preetapan commented:
@diptanu can you update the Changelog for this

jesusvazquez commented:
Hi, I can't find this metrics in here https://www.nomadproject.io/docs/agent/telemetry.html. Can we update the documentation accordingly? This metrics are definitely useful.

schmichael commented:
@jesusvazquez Good call. See #3467
